File: StandardLibrary/runtimes/net/Extern/UUID.cs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0


using Wrappers_Compile;
using ibyteseq = Dafny.ISequence<byte>;
using byteseq = Dafny.Sequence<byte>;
using icharseq = Dafny.ISequence<char>;
using charseq = Dafny.Sequence<char>;

namespace UUID {
    public partial class __default {
        public static _IResult<ibyteseq, icharseq> ToByteArray(icharseq s) {
            try {
                Guid fromString = new Guid(new String(s.CloneAsArray()));
                byte[] bytes = ReOrderUuidBytes(fromString.ToByteArray());
                return Result<ibyteseq, icharseq>.create_Success(byteseq.FromArray(bytes));
            } catch (Exception e) when (e is ArgumentNullException or FormatException or OverflowException) {
                return Result<ibyteseq, icharseq>
                    .create_Failure(Dafny.Sequence<char>.FromString("Unable to convert input to a UUID"));
            }
        }
        public static _IResult<icharseq, icharseq> FromByteArray(ibyteseq b) {
            try {
                byte[] bytes = ReOrderUuidBytes(b.CloneAsArray());
                Guid fromBytes = new Guid(bytes);
                return Result<icharseq, icharseq>.create_Success(charseq.FromString(fromBytes.ToString()));
            }
            catch (Exception e) when (e is ArgumentNullException or ArgumentException) {
                return Result<icharseq, icharseq>.create_Failure(
                    charseq.FromString("Unable to convert bytes to valid UUID"));
            }
        }

        public static _IResult<icharseq, icharseq> GenerateUUID() { 
            Guid uuid = Guid.NewGuid();
            return Result<icharseq, icharseq>.create_Success(charseq.FromString(uuid.ToString()));
        }

        // https://learn.microsoft.com/en-us/dotnet/api/system.guid.tobytearray?redirectedfrom=MSDN&view=net-7.0#remarks
        private static byte[] ReOrderUuidBytes(byte[] uuid) {
            // When using the .ToByteArray() method the byte 
            // representation is different than that of the 
            // string representation. 
            // The order of the beginning four-byte group and
            // the next two two-byte groups is reversed
            // whereas the order of the last two-byte group 
            // and the closing six-byte group is the same 
            byte[] bytes = {
                uuid[3],
                uuid[2],
                uuid[1],
                uuid[0],
                uuid[5],
                uuid[4],
                uuid[7],
                uuid[6],
                uuid[8],
                uuid[9],
                uuid[10],
                uuid[11],
                uuid[12],
                uuid[13],
                uuid[14],
                uuid[15]
            };
            return bytes;
        } 
    }
}

File: AwsEncryptionSDK/runtimes/rust/src/uuid.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

#![deny(warnings, unconditional_panic)]
#![deny(nonstandard_style)]
#![deny(clippy::all)]

use crate::*;
use ::uuid::Uuid;

impl crate::UUID::_default {
    #[allow(non_snake_case)]
    pub fn ToByteArray(
        bytes: &::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
    ) -> ::std::rc::Rc<
        _Wrappers_Compile::Result<
            ::dafny_runtime::Sequence<u8>,
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
        >,
    > {
        let my_str =
            dafny_runtime::dafny_runtime_conversions::unicode_chars_false::dafny_string_to_string(
                bytes,
            );
        match Uuid::parse_str(&my_str) {
                Ok(u) => {
                    let b = u.as_bytes();
                    std::rc::Rc::new(_Wrappers_Compile::Result::Success { value :
                        b.iter().cloned().collect()
                    })
                }
                Err(e) => {
                    std::rc::Rc::new(_Wrappers_Compile::Result::Failure{ error :
                        dafny_runtime::dafny_runtime_conversions::unicode_chars_false::string_to_dafny_string(
                            &format!("{my_str} is not a valid UUID ({e})."))
                    })
                }
            }
    }

    #[allow(non_snake_case)]
    pub fn FromByteArray(
        bytes: &::dafny_runtime::Sequence<u8>,
    ) -> ::std::rc::Rc<
        _Wrappers_Compile::Result<
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
        >,
    > {
        let vec: Vec<u8> = bytes.iter().collect();
        if vec.len() != 16 {
            return std::rc::Rc::new(_Wrappers_Compile::Result::Failure{ error :
                    dafny_runtime::dafny_runtime_conversions::unicode_chars_false::string_to_dafny_string("Not 16 bytes of input to FromByteArray.")
                });
        }
        let bytes: ::uuid::Bytes = vec[..16].try_into().unwrap();
        let uuid = Uuid::from_bytes_ref(&bytes);
        let my_str = uuid.to_string();
        std::rc::Rc::new(_Wrappers_Compile::Result::Success { value :
                dafny_runtime::dafny_runtime_conversions::unicode_chars_false::string_to_dafny_string(&my_str)
            })
    }

    #[allow(non_snake_case)]
    pub fn GenerateUUID() -> ::std::rc::Rc<
        _Wrappers_Compile::Result<
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
            ::dafny_runtime::Sequence<::dafny_runtime::DafnyCharUTF16>,
        >,
    > {
        let my_str = Uuid::new_v4().to_string();
        std::rc::Rc::new(_Wrappers_Compile::Result::Success { value :
                dafny_runtime::dafny_runtime_conversions::unicode_chars_false::string_to_dafny_string(&my_str)
            })
    }
}
